### DISCUSSION TITLE
access self.data.i ?

### DISCUSSION THREAD

**Discussion by traderblakeq** [kernc/backtesting.py#812] [General]
<Data i=86 (2017-08-18 01:30:00) Open=296.19, High=305.55, Low=296.19, Close=305.55, Volume=117.10346, Close_time=2017-08-18 01:45:00, Supertrend=False, Lowerband=277.04990005003623, Upperband=316.8900999499638>

How do I access i=86 

have tried self.data.i --> does not exist  & self.data.index is Timestamp.

2nd question:

I need to slice out from  i=n to i=n

Any suggestion on how I could slice out from self.data - I guess it does not have the .loc

**Comment by kernc** [comment]
> How do I access i=86

`_Data` `__repr__` contains a hint on how "i" is made:
https://github.com/kernc/backtesting.py/blob/65f54f6819cac5f36fd94ebf0377644c62b4ee3d/backtesting/_util.py#L138-L142

Hint: As [`Strategy.data`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.data) is lengthened incrementally, `len(self.data)` should always be the current "i" (index).
 
> I need to slice out from i=n to i=n
> 
> Any suggestion on how I could slice out from self.data - I guess it does not have the .loc

`Strategy.data` doesn't have `.loc`, but individual arrays comprising it have `__getitem__`, so:
```py
i = len(self.data)
cur_value = self.data.Close[i]
values_slice = self.data.Close[i - 5:i]
values_slice_series = self.data.Close.s.iloc[i - 5:i]

**Reply by traderblakeq** [reply]
Thank you!.

Im my usecase I would like to use the numpy array (self.data) for my method, as I guess it is alot faster than Pandas datastructure.

- is it possible to slice from self.data?.

**Reply by kernc** [reply]
If you know any Python, you can just look in the source:
https://github.com/kernc/backtesting.py/blob/65f54f6819cac5f36fd94ebf0377644c62b4ee3d/backtesting/_util.py#L103-L126

Currently, `Strategy.data` `__getitem__` only supports querying for individual columns (i.e. `self.data['Close']`, `self.data.Close`) as ndarrays.

https://github.com/kernc/backtesting.py/blob/65f54f6819cac5f36fd94ebf0377644c62b4ee3d/backtesting/_util.py#L43-L47

This ought to be quite fast indeed!
